script/model/training_util/global_data_constructing_util.py

```diff
@@ -222,16 +224,36 @@ def _predict_grouped_opunit_data(data_list, mini_model_map, model_results_path):
             key = (opunit, x.tobytes())
             if key not in prediction_cache:
                 y_pred = opunit_model.predict(x)
-                # subtract scan from certain double-counted opunits
-                if opunit in data_info.SCAN_SUBSTRACT_UNITS:
-                    scan_y_pred = mini_model_map[OpUnit.SEQ_SCAN].predict(x)
-                    y_pred -= scan_y_pred
                 y_pred = np.clip(y_pred, 0, None)
                 prediction_cache[key] = y_pred
             else:
                 y_pred = prediction_cache[key]
             logging.debug("Predicted {} elapsed time with feature {}: {}".format(opunit_feature[0].name,
                                                                                  x[0], y_pred[0, -1]))
+
+            if opunit in data_info.MEM_ADJUST_OPUNITS:
+                num_tuple = opunit_feature[1][data_info.TUPLE_NUM_INDEX]
+                if opunit == OpUnit.AGG_BUILD:
+                    num_tuple = opunit_feature[1][data_info.CARDINALITY_INDEX]
+
+                pow_high = 2 ** math.ceil(math.log(num_tuple, 2))
+                buffer_size = pow_high * data_info.POINTER_SIZE
+                if opunit == OpUnit.AGG_BUILD and num_tuple <= 256:
+                    buffer_size = 0
+
+                pred_mem = y_pred[0][data_info.TARGET_CSV_INDEX[Target.MEMORY_B]]
+                if pred_mem <= buffer_size:
+                    logging.warning("{} feature {} {} with prediction {} exceeds buffer {}"
+                            .format(data.name, opunit_feature, opunit_feature[1], y_pred[0], buffer_size))
```

Comment:
Also, what is this part doing? Can you add some document for this?

Reply:
Added some documentation for this.